src/cache/impl/LevelDbCache.ts

```diff
@@ -94,21 +94,38 @@ export class LevelDbCache<V = any> implements SortKeyCache<V> {
   }
 
   async put(stateCacheKey: CacheKey, value: V): Promise<void> {
-    const contractCache = this.db.sublevel<string, any>(stateCacheKey.contractTxId, { valueEncoding: 'json' });
+    const contractCache = this.db.sublevel<string, any>(stateCacheKey.key, { valueEncoding: 'json' });
     // manually opening to fix https://github.com/Level/level/issues/221
     await contractCache.open();
     await contractCache.put(stateCacheKey.sortKey, value);
   }
 
-  async delete(contractTxId: string): Promise<void> {
-    const contractCache = this.db.sublevel<string, any>(contractTxId, { valueEncoding: 'json' });
+  async delete(key: string): Promise<void> {
+    const contractCache = this.db.sublevel<string, any>(key, { valueEncoding: 'json' });
     await contractCache.open();
     await contractCache.clear();
   }
 
-  close(): Promise<void> {
-    return null;
-    //return this.db.close();
+  async batch(opStack: BatchDBOp<V>[]) {
+    for (const op of opStack) {
+      if (op.type === 'put') {
+        await this.put(op.key, op.value);
+      } else if (op.type === 'del') {
+        await this.delete(op.key);
+      }
+    }
+  }
+
+  async open(): Promise<void> {
+    if (this.db.status === 'closed' || this.db.status === 'closing') {
```

Comment:
can you please a comment that using 'closing' here is safe - or maybe link the exact code from level library

Reply:
It seems that leveldb is handling all these states pretty good when calling open/close and there seems to be no point in checking it.